tiledb/common/logger.cc

```diff
@@ -328,8 +328,12 @@ std::string global_logger_name(const Logger::Format format) {
 }
 
 Logger& global_logger(Logger::Format format) {
-  static Logger l(global_logger_name(format), Logger::Level::ERR, format, true);
-  return l;
+  // Note: We are *not* deallocating this root `Logger` instance so that
+  // during process exit, threads other than main will not trigger segfault's
+  // if they attempt to log after the main thread has exited.
+  static Logger* l = tiledb_new<Logger>(
```

Comment:
Is the heap profiler ever enabled by the time you get here?  It looks like it is enabled only via C API, not even via environmental option on startup

I suppose if that changes then you will pick it up here which will be nice - unless it expects memory usage to be zero by the time it exits for some reason

Reply:
To my knowledge the heap profiling is a compile time option so “yes” I guess? Assuming that code still works.

Reply:
I would have expected it to be something that has to last for the life of the process but there is a C API `tiledb_heap_profiler_enable` which looks like it connects to the same thing used here

Reply:
Quick skim suggests that the reporter logic is always available, but if its not enabled at compile time it’ll just not have anything to report.

See: https://github.com/TileDB-Inc/TileDB/blob/f4476d523c22e715db05ff6c3807051198c4fb3c/tiledb/common/dynamic_memory/dynamic_memory.h#L91

Reply:
What I'm looking at is:

In `tiledb/common/heap_memory.h`:
```
template <typename T, typename... Args>
T* tiledb_new(const std::string& label, Args&&... args) {
  if (!heap_profiler.enabled()) {
    return new T(std::forward<Args>(args)...);
  }
  ...
}
```

In `tiledb/common/heap_profiler.h`:
```
extern HeapProfiler heap_profiler;

class HeapProfiler {
  inline bool enabled() const {
    // We know that this instance has been initalized
    // if `reserved_memory_` has been set.
    return reserved_memory_ != nullptr;
  }
};
```

And in `tiledb/common/heap_profiler.cc` we have `HeapProfiler::HeapProfiler` initializing its `reserved_memory_` to null.  The enable function I mentioned above initializes this

Reply:
Ah. Looks like that’s a missed optimization issue. The make_shared thing was updated specifically to be compile time switch. You can see the constexpr part here:

https://github.com/TileDB-Inc/TileDB/blob/f4476d523c22e715db05ff6c3807051198c4fb3c/tiledb/common/dynamic_memory/dynamic_memory.h#L213